Preliminary C++ idle timeout implementation

bernardnormier opened the pull request:
This is a preliminary implementation of Xxx.IdleTimeout in C++.

This PR sends heartbeat every idleTimeout / 2 after the previous write.  It doesn't change the ACM code at all.

Note that most of the line additions come from:
 - the new properties
 - the VS project files for the new test

bernardnormier commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -586,11 +594,24 @@ IceInternal::OutgoingConnectionFactory::createConnection(const TransceiverPtr& t
+        if (ci.endpoint->type() == UDPEndpointType)
+        {
+            // No idle timeout or idle check for UDP connections.
+            idleTimeout = chrono::milliseconds(0);
```
I don't think it makes sense to send heartbeats or check for idle with UDP connections.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,91 @@
+    SocketOperation op = _decoratee->write(buf);
+    if (buf.i != start)
+    {
+        // We've sent at least one byte, reschedule the keep alive.
```
I used the tracing code in ConnectionI.cpp to figure out when we write something. I hope this tracing code is complete.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.h:
```diff
@@ -0,0 +1,78 @@
+        {
+        }
+
+        ~IdleTimeoutTransceiverDecorator();
```
Note that Transceiver doesn't have a virtual dtor (or any dtor). Doesn't matter here but that goes against the usual C++ pattern.

bernardnormier commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -49,6 +50,25 @@ namespace
+        {
+            if (auto connection = _connection.lock())
+            {
+                // TODO: handle synchronous and asynchronous failures
```
IceRPC's version:
https://github.com/icerpc/icerpc-csharp/blob/57995b702436169798fc3fe5fd513e6f09591397/src/IceRpc/Internal/IceProtocolConnection.cs#L630

suggests we should do better here. In particular, don't send a heartbeat when one is already outstanding or has failed.

pepone commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,54 @@
+    test(heartbeatCount == 2);
+
+    // Verifies writes (the getHeartbeatCount round-trip) skips a heartbeat.
+    this_thread::sleep_for(chrono::milliseconds(1400));
```
Not sure if I understand this correctly:

We expect 5 heartbeats in 1250 + 1400, but there are just 4 because of the write in the `getHeartbeatCount` invocation.

bernardnormier replied:
Removed this test.

pepone commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2512,6 +2534,36 @@ Ice::ConnectionI::initiateShutdown()
+Ice::ConnectionI::idleCheck() noexcept
+{
+    std::lock_guard lock(_mutex);
+    if (_state == StateActive)
+    {
+        if (_transceiver->hasDataAvailable())
+        {
+            return true;
+        }
+        else
+        {
+            // TODO: replace by ConnectionIdleException.
+            setState(StateClosed, make_exception_ptr(TimeoutException(__FILE__, __LINE__)));
+        }
+    }
+    return false;
```
How does this work? The connection doesn't become idle while there is data available, because we want to give a chance to the thread pool for reading this data.

What is the meaning of the return value?

bernardnormier replied:
Reworked and commented.

pepone commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,185 @@
+SocketOperation
+IdleTimeoutTransceiverDecorator::initialize(Buffer& readBuffer, Buffer& writeBuffer)
+{
+    SocketOperation op = _decoratee->initialize(readBuffer, writeBuffer);
```
We can assert `_heartbeatTimerTask` is already set, to ensure decoratorInit was called before this.

bernardnormier replied:
I don't think this makes sense. We know we always call init immediately after construction, because anybody else has a chance to use this transceiver.

bentoi reviewed (commented):
So if I understand it correctly, the goal is to prevent the connection from timing out if there's data to be read and the thread pool is starved for period which is longer than the read idle timeout (60s by default in IceRPC)?

To me, you should just increase the number of threads in the thread pool to a number of thread that prevents this condition to occur.

The portability `hasBytesAvailable` would also need to be checked.

bernardnormier commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -347,9 +347,17 @@ IceInternal::OutgoingConnectionFactory::OutgoingConnectionFactory(
+        "Ice.IdleTimeout",
+        static_cast<int>(_idleTimeout.count() / 1000)); // in seconds
+    _enableIdleCheck = properties->getPropertyAsIntWithDefault("Ice.EnableIdleCheck", _enableIdleCheck ? 1 : 0) > 0;
+    _idleTimeout = chrono::milliseconds(idleTimeout * 1000);
```
I am not following. idleTimeout is an int here.

pepone replied:
`_idleTimeout.count() / 1000` 

Would become 

` duration_cast<chrono::seconds>(_idleTimeout).count()`


And 

`_idleTimeout = chrono::milliseconds(idleTimeout * 1000);`

Would become:

`_idleTimeout = duration_cast<chrono::milliseconds>(chrono::seconds(idleTimeout));`

bernardnormier replied:
Fixed as suggested.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,185 @@
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator()
+{
+    // It's possible but unlikely init was not called so initialize the timer tasks.
```
I reworked this comment.

bernardnormier commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,54 @@
+
+    p->sleep(2250);
+    test(p->ice_getCachedConnection() == connection); // we still have the same connection
+    test(p->getHeartbeatCount() == 4);
```
I reworked the test. Counting heartbeats is too timing dependent.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,171 @@
+SocketOperation
+IdleTimeoutTransceiverDecorator::write(Buffer& buf)
+{
+    // We're about to write something (maybe asynchronously) - we reschedule the heartbeat as sending a heartbeat now
```
That's what I meant by "write asynchronously" - it may not complete immediately.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,171 @@
+{
+    // We're about to write something (maybe asynchronously) - we reschedule the heartbeat as sending a heartbeat now
+    // or soon would be redundant.
+    rescheduleHeartbeat();
```
The whole point here is the *re* scheduling.
We're about to write something, and are several possibilities:
 a) the write succeeds synchronously
 b) the write writes one or more bytes synchronously; the remaining bytes will be written later typically because the local buffer is full
 c) the write doesn't write anything because the local buffer is full; all the bytes will be written later
 d) the write fails

In none of these cases there is any benefit in sending a concurrent or a slightly delayed heartbeat. We are very much trying to write and a heartbeat is purely about writing when there is no other write activity.

bernardnormier replied:
I updated to logic to cancel the timer task  _before_ the write (for the reason described above) and then schedule it after a successful write.

bentoi replied:
>  We are very much trying to write and a heartbeat is purely about writing when there is no other write activity.

Right but why schedule a heartbeat when writing data is in progress (b or c)? I did expect write to be implemented as follow:
```cpp

SocketOperation
IdleTimeoutTransceiverDecorator::write(Buffer& buf)
{

    _timer->cancel(_heartbeatTimerTask);
    SocketOperation op = _decoratee->write(buf);
    if (op == SocketOperationNone) // or buf.i == buf.b.end()
    {
        // We're done sending data, schedule the sending of a heartbeat.
        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2);
    }
    return op;
}

```

bernardnormier commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2512,6 +2534,41 @@ Ice::ConnectionI::initiateShutdown()
+        {
+            // Schedule or reschedule timer task. Reschedule in the rare case where a concurrent read scheduled the task
+            // already.
+            _timer->schedule(idleCheckTimerTask, idleTimeout, true);
```
Typically, when this function runs, we're going to abort to connection. 
The question is when are we forgiving and don't abort it.

Case 1: the connection is no longer active. In this case, we don't do anything.

Case 2: the connection is waiting to be read, is has not been read yet because we're out of thread. In this case, we reschedule the idleCheck in idle timeout

Case 3: as this idle check is running, another thread is concurrently reading the transceiver and has just (re)scheduled this timer task. The only we can find out of this happened is by canceling the idle timer task (with cancel). When timer->cancel(task) returns true, it means we're in this situation.

And since we've removed the timer task to check if it was there, we need to add it back.

The `if` simply checks "case 2" and "case 3" together, and the action is the same: reschedule timer task.

Other cases: we abort the connection

bernardnormier replied:
> It is not clear to me under what conditions we get "no data available to be read", and cancel returns false

That's the common abort case. The peer didn't send any data for idle time and the idle check has not been rescheduled by a concurrent read.

Keep in mind that timer _removes_ the timer task from its maps just before it executes it. See the comment on TimerTask::cancel.

externl reviewed (approved):
Looks good to me!

bernardnormier commented on cpp/include/IceUtil/Timer.h:
```diff
@@ -145,6 +153,7 @@ namespace IceUtil
         };
 
         void run();
+        bool cancelNoSync(const TimerTaskPtr& task);
```
It's all in the private section of the class

externl replied:
My mistake! I guess I didn't open the code fully in my browser.

bernardnormier commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2155,22 +2156,43 @@ Ice::ConnectionI::create(
-    Ice::ConnectionIPtr conn(
-        new ConnectionI(communicator, instance, monitor, transceiver, connector, endpoint, adapter));
+    shared_ptr<IdleTimeoutTransceiverDecorator> decoratedTransceiver;
+    if (idleTimeout > chrono::milliseconds::zero())
```
I think `zero()` is fine. It's a special static function.

bernardnormier commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,65 @@
+void
+testIdleCheckWithExhaustedThreadPool(const TestIntfPrx& p)
+{
+    cout << "testing idle timeout with exhausted OA thread pool... " << flush;
```
I renamed, updated and added comments to the tests.

pepone commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -347,9 +349,17 @@ IceInternal::OutgoingConnectionFactory::OutgoingConnectionFactory(
     : _communicator(communicator),
       _instance(instance),
       _monitor(new FactoryACMMonitor(instance, instance->clientACM())),
+      _idleTimeout(60000ms),
```
We can also do, which seems easier to read.
```suggestion
      _idleTimeout(60s),
```

pepone commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -347,9 +349,17 @@ IceInternal::OutgoingConnectionFactory::OutgoingConnectionFactory(
+        "Ice.IdleTimeout",
+        static_cast<int>(chrono::duration_cast<chrono::seconds>(_idleTimeout).count()));
+    _enableIdleCheck = properties->getPropertyAsIntWithDefault("Ice.EnableIdleCheck", _enableIdleCheck ? 1 : 0) > 0;
+    _idleTimeout = chrono::duration_cast<chrono::milliseconds>(chrono::seconds(idleTimeout));
```
I suggested this in a previous comment but we can avoid the `duration_cast`. It is not required here because the conversion is not lossy.

```suggestion
    _idleTimeout = chrono::milliseconds(chrono::seconds(idleTimeout));
```

pepone commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -1643,6 +1667,8 @@ IceInternal::IncomingConnectionFactory::IncomingConnectionFactory(
     const shared_ptr<ObjectAdapterI>& adapter)
     : _instance(instance),
       _monitor(new FactoryACMMonitor(instance, dynamic_cast<ObjectAdapterI*>(adapter.get())->getACM())),
+      _idleTimeout(60000ms),
```
```suggestion
      _idleTimeout(60s),
```

pepone commented on cpp/src/Ice/ConnectionFactory.cpp:
```diff
@@ -1695,24 +1721,55 @@ IceInternal::IncomingConnectionFactory::initialize()
+                                       adapterName + ".EnableIdleCheck",
+                                       _enableIdleCheck ? 0 : 1) > 0;
+            }
+            _idleTimeout = chrono::duration_cast<chrono::milliseconds>(chrono::seconds(idleTimeout));
```
```suggestion
            _idleTimeout = chrono::milliseconds(chrono::seconds(idleTimeout));
```

bentoi commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,74 @@
+void
+testIdleCheckDoesNotAbortConnectionWhenThreadPoolIsExhausted(const TestIntfPrx& p)
+{
+    cout << "testing that the idle check does not abort a connection that receives heartbeats... " << flush;
```
It could be less time sensitive to not rely on heartbeats but instead make an asynchronous sleep followed by an ice_ping.

bernardnormier replied:
I don't understand what you're suggesting.

> not rely on heartbeats 

The whole point of this test is to make sure a connection is not idle-check-aborted because it receives heartbeats.

bentoi replied:
Ok

bentoi reviewed (commented):
I wonder if it works with connection "hold". I suspect the held connection might still idle timeout. Unless we remove the hold state, it would be good to check.

bernardnormier commented on cpp/include/IceUtil/Timer.h:
```diff
@@ -43,9 +44,10 @@ namespace IceUtil
         template<class Rep, class Period>
-        void schedule(const TimerTaskPtr& task, const std::chrono::duration<Rep, Period>& delay)
+        void
+        schedule(const TimerTaskPtr& task, const std::chrono::duration<Rep, Period>& delay, bool cancelPrevious = false)
```
Replaced cancelPrevious by new reschedule function.

> We could also get rid of the inlining.

All these functions are inline because they are template functions and therefore must be in the header. You prefer out-of-line template functions in the header?

bentoi replied:
I missed the fact that it was a template, it's fine to keep them in the header.

bernardnormier commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2512,6 +2534,41 @@ Ice::ConnectionI::initiateShutdown()
+Ice::ConnectionI::sendHeartbeat() noexcept
+{
+    lock_guard lock(_mutex);
+    if (_state != StateActive)
```
Fixed as suggested. When we remove ACM, we should naturally merge sendHeartbeat and sendHeatbeatNow.

bernardnormier commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2155,22 +2156,43 @@ Ice::ConnectionI::create(
+        endpoint,
+        adapter));
+
+    if (decoratedTransceiver)
```
That's true. Do you see a better setup?

bentoi replied:
Yes, one that you're not going to like: no transceiver decorator. IMO, it adds more complexity for little gain.

bernardnormier replied:
I definitely prefer the decorator.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,180 @@
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator()
+{
+    // If we destroy this object before calling decoratorInit(), _heartbeatTimerTask and _idleCheckTimerTask will be
```
This would happen with ConnectionI's constructor throws an exception.

Unless we're out of memory, ConnectionI's ctor should not throw. Do you think I should make ConnectionI's ctor noexcept and add an assert in this decorator dtor?

bentoi replied:
We don't handle catostrophic failures such as OOM errors elsewhere so handling it here is a bit surprising. Up to you but I think it's better to not handle it only here.

bernardnormier replied:
Ok, make ConnectionI's ctor noexcept and added assert here.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,180 @@
+SocketOperation
+IdleTimeoutTransceiverDecorator::write(Buffer& buf)
+{
+    // We're about to write something and we don't want to send a concurrent heartbeat.
```
Ok, removed comment.

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    return _decoratee->startWrite(buf);
+
+    // TODO: should we schedule the next heartbeat when the return value is true? Or do we always call finishWrite?
```
Updated startWrite/finishWrite as suggested and added comments.

bernardnormier replied:
> After finishWrite returns, write is called again...

I updated the code (removed the decoration) assuming this is true. 

After transceiver->finishWrite returns, who calls transceiver->write? If it's ConnectionI::write, then something is wrong with our tracing code, as we trace sent bytes after transceiver->finishWrite and in ConnectionI::write.

bernardnormier replied:
The tracing code should be fine, as we measure how much 'buf.i' moves for each call.

bentoi replied:
All readiness notifications from the selector (kqueue, IOCP, etc) are handled by the thread pool leader thread. The leader thread ends up calling one method: the "EventHandler::message" method. So the IOCP completion port returns the write completion for the connection, `Connection::message(SocketOperationWrite)` is called by the thread pool and the implementation of `message` continues the writing (https://github.com/zeroc-ice/ice/blob/05fbaef93520dd07307a8abb7d5143ef935e2e23/cpp/src/Ice/ConnectionI.cpp#L1408)

bernardnormier commented on cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp:
```diff
@@ -0,0 +1,180 @@
+SocketOperation
+IdleTimeoutTransceiverDecorator::read(Buffer& buf)
+{
+    if (_enableIdleCheck)
+    {
+        // We don't want the idle check to run while we're reading, so we reschedule it before reading.
+        _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+    }
+    return _decoratee->read(buf);
```
The read and write logic are not the same:

- for write, we don't want to write more (send a heartbeat) if another write is already in progress, even if this write is very slow / stuck. It would be pointless. That's why we schedule a heartbeat only once the write is completed.
- for read, we want to run the idle check even while a read is in progress, if this read is stuck

So I think the read logic is correct. We get here (in read) only when there is something new to read.

bentoi replied:
> for read, we want to run the idle check even while a read is in progress, if this read is stuck

All the transceiver methods (connect, initialize, read, write) should be block. It's the contract.

bernardnormier replied:
By "stuck" I meant: the next read call is not called in a timely manner. Each read call always completes quickly.

bernardnormier commented on cpp/src/Ice/Network.cpp:
```diff
@@ -1519,6 +1521,21 @@ IceInternal::getRecvBufferSize(SOCKET fd)
+    int rs = ioctl(fd, FIONREAD, &bytesAvailable);
+    assert(rs == 0);
+#endif
+    return rs == 0 && bytesAvailable > 0;
```
If ioctl fails, it's most likely because of a bug in our code, like we're somehow closing the socket while the idle check is running. And I want to get an assert in debug builds when this happens.

In a release build, if it fails, I'd rather not return a false positive. Failure of this call doesn't mean "bytes are available".

bentoi replied:
For other code, we typically through an exception for unexpected failures.

bernardnormier commented on cpp/src/Ice/UdpTransceiver.cpp:
```diff
@@ -469,6 +469,19 @@ IceInternal::UdpTransceiver::finishRead(Buffer& buf)
 #endif
 
+bool
+IceInternal::UdpTransceiver::isWaitingToBeRead() const noexcept
```
It's not.

bernardnormier replied:
Fixed as suggested.

bernardnormier commented on cpp/src/Ice/StreamSocket.cpp:
```diff
@@ -55,6 +55,15 @@ StreamSocket::StreamSocket(const ProtocolInstancePtr& instance, SOCKET fd)
 {
+#if defined(ICE_USE_IOCP)
+    // For the implementation of isWaitingToBeRead.
+    _read.hEvent = WSACreateEvent();
```
> The completion routine is always called once a started read or write completes.

For finishRead: my understanding is the read result is retrieved by a thread from the Ice thread pool which calls GetQueuedCompletionStatus here:

https://github.com/zeroc-ice/ice/blob/59392d7b35720e42a07951beeada22e3a60cfa29/cpp/src/Ice/Selector.cpp#L115

And the whole point of this exercise is to signal "waiting to be read" while the thread pool is out of threads.

Registering a completion routine is another option but I think adding  this event is much simpler. Currently, finishRead is NOT a completion routine.

bentoi replied:
Ok, I understand now.

bernardnormier commented on cpp/src/Ice/Network.cpp:
```diff
@@ -1519,6 +1521,21 @@ IceInternal::getRecvBufferSize(SOCKET fd)
+{
+#ifdef _WIN32
+    unsigned long bytesAvailable = 0;
+    int rs = ioctlsocket(fd, FIONREAD, &bytesAvailable);
```
It is:
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/InputStream.html#available()

bernardnormier commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,74 @@
+    ConnectionPtr connection = p->ice_getCachedConnection();
+    test(connection);
+    p->sleep(2000);                                   // the implementation in the server sleeps for 2,000ms
+    test(p->ice_getCachedConnection() == connection); // we still have the same connection
```
You're right, removed test.

bernardnormier commented on cpp/test/Ice/idleTimeout/AllTests.cpp:
```diff
@@ -0,0 +1,74 @@
+testIdleCheckDoesNotAbortConnectionWhenThreadPoolIsExhausted(const TestIntfPrx& p)
+{
+    cout << "testing that the idle check does not abort a connection that receives heartbeats... " << flush;
+    p->ice_ping();
+    ConnectionPtr connection = p->ice_getCachedConnection();
```
Yes, updated as you suggested.

bernardnormier commented:
> I wonder if it works with connection "hold". I suspect the held connection might still idle timeout. Unless we remove the hold state, it would be good to check.

I added a test (to the Ice/hold test): it works fine - the idle check does not abort a connection when in the holding state, provided this connection is "waiting to be read".

There was a bug though: `Connection::idleCheck` was no-op outside the Active state. I fixed that.

I also added tracing to make sure the idle check actually runs as expected. Here is the server tracing for the new test in Ice/hold:
```
-- 04/11/24 16:01:48183 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: holding tcp connections at 127.0.0.1:12010
-- 04/11/24 16:01:48183 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: sent 25 of 25 bytes via tcp
   local address = 127.0.0.1:12010
   remote address = 127.0.0.1:59620
-- 04/11/24 16:01:48617 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: sent 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:48618 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:48686 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12010
   remote address = 127.0.0.1:59620
-- 04/11/24 16:01:49118 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: sent 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:49119 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:49620 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: sent 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:49621 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:49689 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: the idle check scheduled a new idle check in 1s because the connection is waiting to be read
   local address = 127.0.0.1:12010
   remote address = 127.0.0.1:59620
-- 04/11/24 16:01:50121 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:50122 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: sent 14 of 14 bytes via tcp
   local address = 127.0.0.1:12011
   remote address = 127.0.0.1:59722
-- 04/11/24 16:01:50190 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: accepting tcp connections at 127.0.0.1:12010
-- 04/11/24 16:01:50191 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12010
   remote address = 127.0.0.1:59620
-- 04/11/24 16:01:50191 C:/Users/bernard/source/repos/ice/cpp/test/Ice/hold/msbuild/server/x64/Debug/server.exe: Network: received 14 of 14 bytes via tcp
   local address = 127.0.0.1:12010
   remote address = 127.0.0.1:59620
```

bernardnormier commented on cpp/src/IceIAP/Transceiver.mm:
```diff
@@ -361,6 +361,12 @@ - (void)stream:(NSStream*)stream handleEvent:(NSStreamEvent)eventCode
+bool
+IceObjC::iAPTransceiver::isWaitingToBeRead() const noexcept
+{
+    return [_readStream hasBytesAvailable];
```
I want to keep the IceIAP transport in 3.8.

bernardnormier commented on cpp/src/IceBT/TransceiverI.cpp:
```diff
@@ -96,6 +96,18 @@ IceBT::TransceiverI::read(IceInternal::Buffer& buf)
+        // See comment in read().
+        return true;
+    }
+    return IceInternal::hasBytesAvailable(fd);
```
We should of course test. CI currently builds with BT but does not run any test.

bentoi commented on cpp/src/Ice/ConnectionI.cpp:
```diff
@@ -2535,7 +2592,7 @@ Ice::ConnectionI::sendHeartbeatNow()
             sendMessage(message);
         }
-        catch (const LocalException&)
+        catch (...)
```
Throughout the code, we only catch expected exceptions and let unexpected ones go through. Here, unexpected exceptions are catch at a higher level in ACM.cpp where the catch handler endup printing an error with the logger. 

If you want to change this pattern, it should be changed throughout the code.

bernardnormier replied:
I reverted this update. Not essential for this PR.

bentoi commented on cpp/src/Ice/StreamSocket.cpp:
```diff
@@ -321,6 +336,24 @@ StreamSocket::write(const char* buf, size_t length)
+StreamSocket::isWaitingToBeRead() const noexcept
+{
+#if defined(ICE_USE_IOCP)
+    if (hasBytesAvailable(_fd))
```
Most likely because there's already read data buffered and there's no need to start an asynchronous read on the underlying tcp/ssl transceiver. That's the case when `preRead` returns false here https://github.com/zeroc-ice/ice/blob/05fbaef93520dd07307a8abb7d5143ef935e2e23/cpp/src/Ice/WSTransceiver.cpp#L702

In this case, we just trigger a read event on the completion port. In turn, the thread pool will call right away `finishRead`.

bentoi reviewed (approved):
I don't like much this fix.

I would have considered the alternative of implementing MaxDispatch/MaxConnections instead. And consider of getting of client/server/OA thread pools but instead do like .NET: a single thread pool with no max (or a very large maximum).
